core/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -2058,6 +2058,12 @@ public void acquireReplicaOperationPermit(final long operationPrimaryTerm, final
                                 "shard term already update.  op term [" + operationPrimaryTerm + "], shardTerm [" + primaryTerm + "]";
                             primaryTerm = operationPrimaryTerm;
                             updateGlobalCheckpointOnReplica(globalCheckpoint);
+                            logger.trace(
+                                    "detected new primary with primary term [{}], resetting local checkpoint from [{}] to [{}]",
+                                    operationPrimaryTerm,
+                                    getLocalCheckpoint(),
+                                    getGlobalCheckpoint());
+                            getEngine().seqNoService().resetLocalCheckpoint(getGlobalCheckpoint());
```

Comment:
As we are feeding this method the current global checkpoint, which could be still unknown, is it possible that we call `resetLocalCheckpoint` with `SequenceNumbersService.UNASSIGNED_SEQ_NO`? If so, I think that that would be bad. The method `resetLocalCheckpoint` should have an assertion, similar to the constructor. Also we need to make sure to special-case this.